Improve display address handling for external signer

Sjors opened the pull request:
* HWI returns the requested address: as a sanity check, we now compare that to what we expected
   * external signer documentation now reflects that HWI alternatives must implement this check
* both RPC and GUI will now return an error text, rather than just fail (the GUI even failed silently in some cases)

Sjors committed:
wallet: compare address returned by displayaddress

Update external signer documentation to reflect this requirement, which HWI already implements.

jonatack commented on test/functional/wallet_signer.py:
```diff
@@ -111,8 +111,10 @@ def test_valid_signer(self):
-        result = hww.walletdisplayaddress(address1)
-        assert_equal(result, {"address": address1})
+        for address in [address1, address2, address3]:
+            print(address)
```
c83ab3059 The printed addresses correspond to the updated mocks. I think you intended to remove this line, though.

jonatack commented on src/wallet/external_signer_scriptpubkeyman.h:
```diff
@@ -27,7 +29,11 @@ class ExternalSignerScriptPubKeyMan : public DescriptorScriptPubKeyMan
+  * Display address on the device and verify that the returned value matches.
+  * @returns an error message or nullopt
+  */
+  std::optional<bilingual_str> DisplayAddress(const CTxDestination& dest, const ExternalSigner &signer) const;
```
c83ab3059 since you're touching this line
```suggestion
  std::optional<bilingual_str> DisplayAddress(const CTxDestination& dest, const ExternalSigner& signer) const;
```

jonatack commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -49,15 +51,26 @@ ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
-    signer.DisplayAddress(descriptor->ToString());
-    // TODO inspect result
-    return true;
+    UniValue result = signer.DisplayAddress(descriptor->ToString());
```
c83ab3059 
```diff
@@ -54,11 +54,11 @@ ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
 std::optional<bilingual_str> ExternalSignerScriptPubKeyMan::DisplayAddress(const CTxDestination& dest, const ExternalSigner &signer) const
 {
     // TODO: avoid the need to infer a descriptor from inside a descriptor wallet
-    CScript scriptPubKey = GetScriptForDestination(dest);
+    const CScript& scriptPubKey = GetScriptForDestination(dest);
     auto provider = GetSolvingProvider(scriptPubKey);
     auto descriptor = InferDescriptor(scriptPubKey, *provider);
 
-    UniValue result = signer.DisplayAddress(descriptor->ToString());
+    const UniValue& result = signer.DisplayAddress(descriptor->ToString());
```

jonatack commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -49,15 +51,26 @@ ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
+
+    if (ret_address.getValStr() != EncodeDestination(dest)) {
+        return strprintf(_("Signer echoed unexpected address %s"), ret_address.getValStr());
+    }
```
Functional tests of these errors would be nice, if feasible.

Sjors replied:
This particular one should be doable.

Sjors replied:
Added.

jonatack reviewed (commented):
Approach ACK

luke-jr commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -49,15 +51,26 @@ ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
+    const UniValue& error = find_value(result, "error");
+    if (error.isStr()) return strprintf(_("Signer returned error: %s"), error.getValStr());
+
+    const UniValue& ret_address = find_value(result, "address");
+    if (!ret_address.isStr()) return _("Signer did not echo address");
```
Won't this break older external signers?

Sjors replied:
HWI has always done this, we just didn't check. That said, it's not (yet) in the spec (which may be a bit outdated) so other signer software might not do this. Do you know of any? cc @achow101 

 https://github.com/bitcoin/bitcoin/blob/master/doc/external-signer.md#displayaddress-optional

achow101 replied:
AFAIK there aren't any other external signer software yet.

luke-jr commented on src/qt/walletmodel.cpp:
```diff
@@ -560,13 +560,16 @@ bool WalletModel::bumpFee(uint256 hash, uint256& new_hash)
+    if (error) {
+        QMessageBox::warning(nullptr, tr("Signer error"), QString::fromStdString(error->translated));
+    }
+    return true;
```
```suggestion
    return !error;
```

Sjors commented:
Rebased and addressed comments.

I could change "Signer did not echo address" from an error to a warning. However I'd rather only do this is we know there is an HWI clone out there that relies on the (less safe) earlier behavior. I updated the documentation. See also https://github.com/bitcoin/bitcoin/pull/24313#discussion_r808915455.

Sjors commented:
Rebased after #23578.

Sjors commented:
Rebased and tested that it still works.

Sjors commented:
Rebased because `<key_io.h>` moved (Kernel related?) and to add an explicit `include <univalue.h>` (macOS / clang doesn't care, but CI does and it makes sense anyway).

Sjors commented:
Oops, I didn't rebase all the way. Specifically not past #27605. Will fix.

Sjors commented:
Ah yes, broken my own #26076.

Sjors committed:
test: use h marker for external signer mock

Consistent with #26076

Sjors commented:
CI is happy now

achow101 commented on src/interfaces/wallet.h:
```diff
@@ -125,7 +125,7 @@ class Wallet
 
     //! Display address on external signer
-    virtual bool displayAddress(const CTxDestination& dest) = 0;
+    virtual std::optional<bilingual_str> displayAddress(const CTxDestination& dest) = 0;
```
In 8218ce6b02b24eeef31f4e343ea777b7844907fc "wallet: return and display signer error"

Returning the error in an optional feels a bit wrong since it inverts are typical pattern. Rather than that the result is equivalent to `true`, we have to check that it's equivalent to `false`.

I think it would be better to use `util::Result<bool>` instead of an optional in order to preserve our typical error checking pattern.

Sjors replied:
Done

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/24313#discussion_r1203035210

> Returning the error in an optional feels a bit wrong since it inverts the typical pattern. Rather than checking that the result is equivalent to `true`, we have to check that it's equivalent to `false`.

Returning `optional<bilingual_str>` or `Result<bool>` both seem ok here, but the ideal thing would be to return `Result<void>` so there are just 2 cases for calling code to handle 1-success, and 2-failure with an error message. Not 3 cases to handle: 1-success returning true, 2-success returning false, and 3-failure with an error message.

Complaining about the `optional<bilingual_str>` pattern is a recurring topic, see https://github.com/bitcoin/bitcoin/pull/27632#discussion_r1192007516, https://github.com/bitcoin/bitcoin/pull/27632#discussion_r1194858242, https://github.com/bitcoin/bitcoin/pull/25648#discussion_r935978206 and I have a PR to replace uses of `optional<bilingual_str>` with `util::Result` in #25977, but it's blocked waiting on a complicated base PR which adds support for `Result<void>`

Since this keeps coming up, though I think I will just add simple support for `Result<void>` to #25977 and decouple it from the base PR so it can get unstuck.

Just thinking out loud here, though (and advertising #25977)

Sjors replied:
I switched to using util::Result in my last rebase a few months ago.

Sjors commented:
Rebased and switched to `Result<void>`.

ryanofsky commented on src/wallet/rpc/addresses.cpp:
```diff
@@ -780,9 +780,8 @@ RPCHelpMan walletdisplayaddress()
-                throw JSONRPCError(RPC_MISC_ERROR, "Failed to display address");
-            }
+            util::Result<void> res = pwallet->DisplayAddress(dest);
+            if (!res) throw JSONRPCError(RPC_MISC_ERROR, util::ErrorString(res).translated);
```
In commit "wallet: return and display signer error" (0e0163a062b09a3f762d7dd0af3d6e78e675aae1)

s/translated/original/ here since RPC methods return untranslated error strings

achow101 commented:
re-ACK 5536ace894e67322161a649f998265f461fbc8ad

brunoerg commented:
Concept ACK

brunoerg commented on src/wallet/external_signer_scriptpubkeyman.cpp:
```diff
@@ -52,15 +53,19 @@ ExternalSigner ExternalSignerScriptPubKeyMan::GetExternalSigner() {
-bool ExternalSignerScriptPubKeyMan::DisplayAddress(const CScript scriptPubKey, const ExternalSigner &signer) const
+bool ExternalSignerScriptPubKeyMan::DisplayAddress(const CTxDestination& dest, const ExternalSigner &signer) const
 {
     // TODO: avoid the need to infer a descriptor from inside a descriptor wallet
```
Do you intend to address this TODO?

Sjors replied:
Not in this PR, unless someone else gives me the code for it :-)

brunoerg commented on src/wallet/wallet.cpp:
```diff
@@ -2644,7 +2644,7 @@ bool CWallet::DisplayAddress(const CTxDestination& dest)
         return signer_spk_man->DisplayAddress(dest, signer);
     }
-    return false;
+    return util::Error{_("No external signer ScriptPubKeyManager")};
```
In 5536ace894e67322161a649f998265f461fbc8ad: I think we can improve the "No external signer ScriptPubKeyManager" message, it doesn't sound intuitive. First time I tested it prior reading the code, I got this error and I thought I didn't have any external signer, so I checked it with `enumerate`.

Perhaps: "There is no ScriptPubKeyManager for this address".

Sjors replied:
Will consider new wording if I need to touch / rebase.

Sjors replied:
Done

brunoerg commented:
I tested it with HWI and Ledger Nano S Plus. 

--------

- I created a wallet:
`./src/bitcoin-cli -named createwallet wallet_name=ledger3 disable_private_keys=true descriptors=true external_signer=true`

- Got a new address:
```sh
$ ./src/bitcoin-cli -rpcwallet=ledger3 getnewaddress
bc1qfshzjl770c6qz560cade6ynswj2a3scqzm4a4m
```
- Ran `walletdisplayaddress`:
```sh
$ ./src/bitcoin-cli -rpcwallet=ledger3 walletdisplayaddress "bc1qfshzjl770c6qz560cade6ynswj2a3scqzm4a4m"
```
- Checked it in the Ledger device:
<img src="https://github.com/bitcoin/bitcoin/assets/19480819/88003e4f-e72e-4c74-89ef-3c0831e5b545" data-canonical-src="https://github.com/bitcoin/bitcoin/assets/19480819/88003e4f-e72e-4c74-89ef-3c0831e5b545" width="200" height="150" />


- I locked the Ledger device and tried `walletdisplayaddress` to check the error message:

This PR:
`'/Users/bg/projects/HWI/hwi.py' error: Could not open client or get fingerprint information: Exception: invalid status 0x6982`

Master:
`'/Users/brunogarcia/projects/HWI/hwi.py' error: Could not open client or get fingerprint information: Exception: invalid status 0x6982`

-------------------------

- Ran `walletdisplayaddress` with a random address:

On master:
```
error code: -1
error message:
Failed to display address
```

This PR:
```
error code: -1
error message:
No external signer ScriptPubKeyManager
```

Sjors commented:
Mmm, on the bright side, both the master and this PR have a really unclear message when the device is not connected :-)

Sjors commented:
@achow101 any thoughts on having HWI catch this cryptic `0x6982` status code and throw a more clear error (that we can then just pass on to the user)?

brunoerg commented:
> @achow101 any thoughts on having HWI catch this cryptic 0x6982 status code and throw a more clear error (that we can then just pass on to the user)?

Not sure, but this status code is specific from Ledger?

Sjors commented:
@brunoerg most likely yes, but I can't find a full list anywhere.

brunoerg commented:
> @brunoerg most likely yes, but I can't find a full list anywhere.

Me either. I've some of them but I manually mapped.

Sjors committed:
wallet: return and display signer error

Both RPC and GUI now render a useful error message instead of (silently) failing.

Replace bool with util::Result<void> to clarify that this either succeeds or returns an error message.

Sjors commented:
Rebased to give CI another try.

HWI now returns better errors, though I have not yet tested it here.

Sjors commented:
Tested with HWI 2.4.0

When a Ledger Nano X is in screen saver mode, I now get the following error:

```
'...hwi' error: Could not open client or get fingerprint information: ('0x5515', 'Error in <DefaultInsType.GET_VERSION: 1> command', '')
```

That's what HWI returns, so can't make it much better...

When the Bitcoin app is not opened on the device the error is more clear:

```
'...hwi' error: Could not open client or get fingerprint information: Ledger is not in either the Bitcoin or Bitcoin Testnet app
```

Sjors commented:
Rebased after #28981 (I previously already tested on top of that).

Briefly tested with HWI 3.0 as well on a Ledger Nano X.

brunoerg commented:
> Tested with HWI 2.4.0
> 
> When a Ledger Nano X is in screen saver mode, I now get the following error:
> 
> ```
> '...hwi' error: Could not open client or get fingerprint information: ('0x5515', 'Error in <DefaultInsType.GET_VERSION: 1> command', '')
> ```
> 
> That's what HWI returns, so can't make it much better...
> 
> When the Bitcoin app is not opened on the device the error is more clear:
> 
> ```
> '...hwi' error: Could not open client or get fingerprint information: Ledger is not in either the Bitcoin or Bitcoin Testnet app
> ```

I just did a brief test and got same result for master branch, didn't test from this PR yet.

brunoerg commented:
I did same test as https://github.com/bitcoin/bitcoin/pull/24313#issuecomment-1867691764 but with the latest version of HWI. 

```sh
 ./src/bitcoin-cli -rpcwallet=ledger7 walletdisplayaddress "bc1qfshzjl770c6qz560cade6ynswj2a3scqzm4a4m"
error code: -1
error message:
'/Users/brunogarcia/projects/HWI/hwi.py' error: Could not open client or get fingerprint information: Ledger is not in either the Bitcoin or Bitcoin Testnet app
```

```sh
 ./src/bitcoin-cli -rpcwallet=ledger7 walletdisplayaddress "randomaddresshere"
error code: -1
error message:
There is no ScriptPubKeyManager for this address
```

achow101 commented:
ACK 4357158c4712d479522d5cd441ad4dd1693fdd05
